subprojects/docs/src/docs/userguide/lazyConfiguration.adoc

```diff
@@ -15,40 +15,226 @@
+
+Lazy properties are intended to be passed around and only evaluated when required (usually, during the execution phase). For more information about the Gradle build phases, please see <<sec:build_phases>>.
+
+The following demonstrates a task with a read-only property and a configurable property:
+
+++++
+<sample id="usePropertyAndProvider" dir="providers/propertyAndProvider" title="Using a read-only and configurable property">
+    <sourcefile file="build.gradle"/>
+    <output args="greeting"/>
+</sample>
+++++
+
+The `Greeting` task has a `Property<String>` for the mutable part of the message and a `Provider` for the modified, read-only, message.
+
+[NOTE]
+====
+
+Note that Groovy Gradle DSL will generate setter methods for each `Property`-typed property in a task implementation. These setter methods allow you to configure the property using the assignment (`=`) operator as a convenience.
+
+====
+
+[[sec:why_new_types]]
+=== Why new types?
+The Provider API is a public implementation of a feature Gradle has had internally for a long time--convention mapping. Provider types are faster, more understandable and better instrumented than the internal convention mapping mechanisms. The Provider API also enables more consistent build logic since it allows a plugin or task to express the relationship between properties without worrying about when particular values are determined.
+
+[[sec:creating_property_provider]]
+=== Creating a Property or Provider
+
+If provider types are not intended to be implemented directly by build script or plugin authors, how do you create new one? Gradle provides various factory APIs to create new instances of both api:org.gradle.api.provider.Provider[] and api:org.gradle.api.provider.Property[]:
+
+- api:org.gradle.api.provider.ProviderFactory#provider(java.util.concurrent.Callable)[] instantiates a new `Provider`. A handy method is available at api:org.gradle.api.Project#provider(java.util.concurrent.Callable)[]. If a api:org.gradle.api.Project[] instance isn’t available, you can inject api:org.gradle.api.provider.ProviderFactory[] through a constructor or method.
+- api:org.gradle.api.model.ObjectFactory#property(java.lang.Class)[] instantiates a new `Property`. An instance of the api:org.gradle.api.model.ObjectFactory[] can be referenced from api:org.gradle.api.Project#getObjects()[] or by injecting ObjectFactory through a constructor or method.
+
+[NOTE]
+====
+
+api:org.gradle.api.Project[] does not provide a specific method signature for creating a provider from a `groovy.lang.Closure`. When writing a plugin with Groovy, you can use the method signature accepting a `java.util.concurrent.Callable` parameter. Groovy's http://docs.groovy-lang.org/next/html/documentation/core-semantics.html#_assigning_a_closure_to_a_sam_type[Closure to type coercion] will take of the rest.
```

Comment:
Maybe also explain what to do in Kotlin.

Reply:
I think we're going to leave Kotlin DSL out of this for now.  I've opened a separate issue for kotlin-dsl to look at the ergonomics of Property/Provider and update their samples.